### ISSUE TITLE
Error building kubectl

### ISSUE THREAD

**Issue by jibanes** [afbjorklund/buildroot4kubernetes#17]
```
>>> kubernetes-kubectl 1.23.0 Building
echo kubernetes-kubectl-1.23.0
kubernetes-kubectl-1.23.0
PATH="/home/xxx/buildroot/buildroot/output/host/bin:/home/xxx/buildroot/buildroot/output/host/sbin:/bin:/usr/bin:/usr/local/go/bin" /usr/bin/make -j9 -C /home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0 generated_files
make[2]: Entering directory '/home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0'
make[3]: Entering directory '/home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0'
+++ [0306 09:36:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
+++ [0306 09:36:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
> non-static build: k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
> non-static build: k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/deepcopy-gen
+++ [0306 09:36:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [0306 09:36:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [0306 09:36:01] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
> non-static build: k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/defaulter-gen
> non-static build: k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/conversion-gen
> non-static build: k8s.io/kubernetes/./vendor/k8s.io/kube-openapi/cmd/openapi-gen
ln: failed to create symbolic link '/home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0/_output/bin': File exists
!!! [0306 09:36:11] Call tree:
!!! [0306 09:36:11]  1: hack/make-rules/build.sh:28 kube::golang::place_bins(...)
make[3]: *** [Makefile.generated_files:402: _output/bin/defaulter-gen] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: Leaving directory '/home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0'
make[2]: *** [Makefile:528: generated_files] Error 2
make[2]: Leaving directory '/home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0'
make[1]: *** [package/pkg-generic.mk:295: /home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0/.stamp_built] Error 2
make[1]: Leaving directory '/home/xxx/buildroot/buildroot'
make: *** [makefile:47: output/disk.img] Error 2
$ ls -l /home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0/_output/bin
lrwxrwxrwx 1 xxx xxx 125 Mar  6 09:36 /home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0/_output/bin -> /home/xxx/buildroot/buildroot/output/build/kubernetes-kubectl-1.23.0/_output/local/bin/linux/amd64
```

Do you know what could cause this?
Could it be that kubectl is built more than once? Or the target generated_files invoked more than once?
If I run the make again, without a make clean, it finishes the build and generates the image.

**Comment by jibanes** [comment]
```
kube::golang::place_bins() {
  local host_platform
  host_platform=$(kube::golang::host_platform)

  V=2 kube::log::status "Placing binaries"

  local platform
  for platform in "${KUBE_CLIENT_PLATFORMS[@]}"; do
    # The substitution on platform_src below will replace all slashes with
    # underscores.  It'll transform darwin/amd64 -> darwin_amd64.
    local platform_src="/${platform//\//_}"
    if [[ "${platform}" == "${host_platform}" ]]; then
      platform_src=""
      rm -f "${THIS_PLATFORM_BIN}"
      ln -s "${KUBE_OUTPUT_BINPATH}/${platform}" "${THIS_PLATFORM_BIN}"
    fi

    local full_binpath_src="${KUBE_GOPATH}/bin${platform_src}"
    if [[ -d "${full_binpath_src}" ]]; then
      mkdir -p "${KUBE_OUTPUT_BINPATH}/${platform}"
      find "${full_binpath_src}" -maxdepth 1 -type f -exec \
        rsync -pc {} "${KUBE_OUTPUT_BINPATH}/${platform}" \;
    fi
  done
}
```

**Comment by afbjorklund** [comment]
Incremental builds do not always work, so I normally try a "clean" first

**Comment by jibanes** [comment]
looks like kubernetes-kubectl.mk, and kubernetes_kubeadm.mk both do a make generated_files in the same $(KUBERNETES_DL_SUBDIR), isn't that what causes the link to be already there?

**Comment by afbjorklund** [comment]
Hmm, they were not supposed to change the download directory but generate the files in the build directory.

**Comment by jibanes** [comment]
I've changed your mk file to this, deleted my buildroot and rebuilt it, works; but I still don't know the root cause.
At any rate this change shouldn't break anything.
```
################################################################################
#
# kubernetes kubectl
#
################################################################################

KUBERNETES_KUBECTL_VERSION = 1.23.0
KUBERNETES_KUBECTL_SITE = $(KUBERNETES_SITE)
KUBERNETES_KUBECTL_DL_SUBDIR = $(KUBERNETES_DL_SUBDIR)
KUBERNETES_KUBECTL_LICENSE = Apache-2.0
KUBERNETES_KUBECTL_LICENSE_FILES = LICENSE

KUBERNETES_KUBECTL_SOURCE = $(KUBERNETES_SOURCE)

define KUBERNETES_KUBECTL_BUILD_CMDS
        echo $(KUBERNETES_KUBECTL_BASENAME)
        -$(TARGET_MAKE_ENV) $(MAKE) -C $(@D) generated_files
        $(TARGET_MAKE_ENV) $(MAKE) -C $(@D) all WHAT="cmd/kubectl" KUBE_BUILD_PLATFORMS="linux/$(GO_GOARCH)"
endef

define KUBERNETES_KUBECTL_INSTALL_TARGET_CMDS
        $(INSTALL) -D -m 0755 $(@D)/_output/local/bin/linux/$(GO_GOARCH)/kubectl $(TARGET_DIR)/usr/bin
endef

$(eval $(generic-package))
```

**Comment by jibanes** [comment]
It built successfully with the hack above. So I deleted buildroot again, and rebuilt from scratch, and now I hit the same issue with kubernetes-kubelet this time. I really don't know what it could be though.

**Comment by afbjorklund** [comment]
Ok, thanks for reporting

**Comment by jibanes** [comment]
It's worth noting that the minus I have added to the three binaries still produces them well, it just makes buildroot run consistently, but I'm all ears if you find the root cause though; please note that I am building buildroot on nfs, so it's not the fastest (it's actually fairly slow), but it locks properly, so I'm not certain this could be an issue, but worth mentioning here.

**Comment by afbjorklund** [comment]
The files are generated in the build directory, not in the download (dl) directory which is **only** the tarballs.

So I don't know why there should be interference between different build directories, when generating source ?

**Comment by afbjorklund** [comment]
The real problem here seems to be that the Kubernetes make files are not concurrency-safe (need -j1).

It doesn't have the proper dependencies set up, so it is not safe to run the build with `make -j9` or such.

Fix: 04e18e59c75f29facd939fbe101ca40d196d9331

Possibly the slow network disk used made the race window bigger and thus failing the build more often ?

**Comment by jibanes** [comment]
Thank you
